kytos/core/interface.py

```diff
@@ -173,19 +177,295 @@ def status_reason(self):
             super().status_reason
         )
 
-    def set_available_tags(self, iterable):
+    @staticmethod
+    def range_intersection(
+        ranges_a: list[list[int]],
+        ranges_b: list[list[int]]
+    ) -> list[list[int]]:
+        """Returns the intersection between two list
+        of ranges"""
+        result = []
+        a_i, b_i = 0, 0
+
+        while a_i < len(ranges_a) and b_i < len(ranges_b):
+            fst_a, snd_a = ranges_a[a_i]
+            fst_b, snd_b = ranges_b[b_i]
+
+            # Moving forward with non-intersection
+            if snd_a < fst_b:
+                a_i += 1
+            elif snd_b < fst_a:
+                b_i += 1
+            else:
+                # Intersection
+                intersection_start = max(fst_a, fst_b)
+                intersection_end = min(snd_a, snd_b)
+                result.append([intersection_start, intersection_end])
+
+                if snd_a < snd_b:
+                    a_i += 1
+                else:
+                    b_i += 1
+        return result
+
+    def get_available_tags(self, tag_type: str = '1') -> list[list[int]]:
+        """Get available tags, if there is tag_ranges restriction
+        self.available_tags are not truly available"""
```

Comment:
I have an impression that's it's deviating from the original intention and purpose of `available_tags`, even when evolving it it's supposed to keep track of only what's actually still available. But, then when a new smaller subset tag_ranges is set, then it should remove from `available_tags` too, and then when `tag_ranges` is expanded again, then the new difference should be included in the `available_tags` again (since we know that during an expansion of the range there's no concern of that's being used, but whatever extra diff is added it also needs to be in available_tags). 

It's an interesting situation, we'll need to better define and sort this out here. Not maintaining `available_tags` as actually what's indeed available will make it harder when getting the next available tag, and hard to reason about, I'll encourage you to review and think about this part here.

Reply:
Yeah, thought it was going to be easier to maintain but It got more complicated later. Now `available_tags` is resized every time `tag_ranges` changes.